Validation Zones (#862)
cardano-foundation/CIPs: CIP-0118/README.md

---
CIP: 0118
Title: Babel fees via validation zones
Category: Ledger
Status: Proposed
Authors:
    - Polina Vinogradova <polina.vinogradova@iohk.io>
    - Michael Peyton Jones <michael.peyton-jones@iohk.io>    
    - Will Gould <will.gould@iohk.io>
Implementors: []
Discussions:

>     - https://github.com/cardano-foundation/CIPs/pull/780

Reviewer: ```suggestion
    - https://github.com/cardano-foundation/CIPs/pull/780
    - https://github.com/cardano-foundation/CIPs/pull/862
```

is for the Dapp to cover the fees and min-UTxO value without being compensated in
other tokens (although the Dapp itself may require depositing assets into it).

### Bridges

Centralized approaches to bridging tokens across blockchains have been developed,
e.g. by [SingularityNET](https://singularitynet.gitbook.io/welcome-to-singularitynet/bridge/overview).
A decentralized approach is much more difficult due to the requirement that
there must be some way to verify data about the operation of an external piece of
distributed software (on the side of both chains involved).
This CIP makes the centralized version of the bridging process slightly more
streamlined on the side of Cardano. Let B be a _trusted_ bridge. The untrusted
case requires additional research.

1. Party A sends an transaction T1 disposing of some tokens on chain D (either by burning
  them or putting them in a special token pool) to B as well as to chain D
2. Party A sends a transaction T2 with a request for the same amount of
  tokens as in (1) on Cardano to B
3. Bridge B waits for T1 to settle on D, then submits the zone `[T2; T3]` to
  Cardano, where `T3` fulfills the request of `T2`

> ## Towards Better Designs

Reviewer: ```suggestion
### Towards Better Designs
```
(with header level changes all the way down from here, and maybe part of the way up)

@polinavino this material is important & exactly in keeping with our goals for _Rationale_... to compare & contrast with other potential designs.  To put it in an unconventional (and therefore unexpected) top level section would likely be a distraction to reviewers and potential implementors.  For readers it does make sense to keep it in a separate section, but the section title & order it appears in the source material is sufficient here to categorise it as future rather than current consideration without making it a top-level heading.

Supporting these sorts of intents require the definition of new types of temporary ledger
state structures, as well as changes to zone-level constraints.
We believe that this is a fruitful direction that we can extend in future, and,
as such, we think we can progress with this design even if we don’t have a full
picture of where we might want to go with intents, although this is a significant risk.
However, at this time there is no proposal for a general solution, i.e. one where introducing
the ability to process each new kind of intent does not require specific ledger
changes. Instead, the ideal intents-processing solution using validation zones
would support intents whose zone-level constraints and any temporary data that
must be recorded are specified _within the zone itself_, providing much more
flexibility in the kinds of intents that are possible.

It may be that the best way to implement intents on Cardano is entirely different from
the underspecified transactions and validation zones approach we are taking here.
In that case such a design might entirely obsolete this one, making it an expensive
and pointless burden to maintain in future.

## Path to Active

> ### Software Readiness Level

Reviewer: For consistency with other CIPs (mainly for review in parallel with 100+ others) this section needs to be broken into _Acceptance_ and _Implementation_ ... I guess since it refers to testing functionality then it would be on the Implementation path.

When done sifting material around in this section it will also help for these items to be GitHub formatted tickboxes (`- [ ]`) but I am not going to be pedantic about it especially at this stage.

Author: Sounds good, will do that later today!

---

> CIP: 0118

Reviewer 2: ```suggestion
CIP: 118
```
leading `0`s not required for the header

> Title: Babel fees via validation zones

Reviewer 2: Do we want to align the title of the pull request with this title?

Reviewer: Yes... especially since as I understand it this proposal will enable applications of far wider range than just Babel Fees (@polinavino?).  The very general title should be fine for this CIP because it doesn't look like a "competing" proposal for Validation Zones would have such a good chance as this one of getting endorsement from the Ledger team.

Author: that make sense (will change it)!

The only fields updated are :

1. the collateral inputs are removed from the UTxO (and change returned)
2. `fees` is updated to `fees` + the collateral amount in the last transaction of the zone

#### UTXO

The `UTXO` rule is updated to additionally check that the `fulfills`
of the transaction being processed are present in the `FRxO` set.
The `UTXO` rule checks the preservation of value (POV), which compares the `produced`
and `consumed` value for the given transaction. The POV calculation is updated by
adding

1. the sum total value in the transaction `fulfills` to the `produced` value
2. the sum total value in the transaction `requests` to the `consumed` value

#### UTXOS

The `UTXOS` rule is the one that actually computes the updates to
the UTxO state done by a transaction. The UTxO set update remains the same
as in the existing design. The `FRxO` is updated by removing
all entries that are indexed by the values in the `fulfills` of the
transaction, then adding all entries of the form `(txID, ix) ↦ r`, where
`txID` is the ID of the transaction being processed, and `ix ↦ r` is
in the `requests` of the transaction.

> ## System Component Changes

Reviewer 2: ```suggestion
### System Component Changes
```

Reviewer: Similar justification for this as for https://github.com/cardano-foundation/CIPs/pull/862#discussion_r1688272866.

are valid outside a zone. We conjecture that infrastructure we propose can be used for processing
other types of underspecified transactions, and therefore more sophisticated
intents. A proof of this requires a
fleshed-out design for the expanded intent-processing system, which is outside the
scope of this CIP. However, possibilities include :

- Intent transactions that allow output references to be removed and replaced with their corresponding
output data, while intent-settling transactions provide the matching output references
for those outputs to fulfill this intent. This is the ledger component of a proposal
for a light client solution.

- Intents that support more advanced kinds of swaps, such as limit or market orders,
and do not require the entire quantity offered to be swapped at once.

Supporting these sorts of intents require the definition of new types of temporary ledger
state structures, as well as changes to zone-level constraints.
We believe that this is a fruitful direction that we can extend in future, and,
as such, we think we can progress with this design even if we don’t have a full

> picture of where we might want to go with intents, although this is a significant risk.

Reviewer 3: @polinavino I really appreciate the R&D!

IIUC my biggest issue is the absence of partial fulfillment (ie, the values must match exactly). I think the absence of this feature severely minimizes its economic utility since the double coincidence of wants is likely to be a huge obstacle for most economic applications that try to use these Validation Zones.

> ... we think we can progress with this design even if we don't have a full picture of where we might want to go with intents, although this is a significant risk.

I don't agree with this stance. In my view, the minimum viable product of an intent system *must* support partial fulfillment. I am definitely no where near as well versed with the Validation Zone design as you, but it is not clear to me that the current design actually *can* be cleanly extended to support partial fulfillment. I personally don't find the other use cases compelling enough to take on this risk since they seem too niche. Even the DEX aggregator example only really seems relevant for liquidity pool based DEXs.

I would totally be in favor of this if I had more assurances that Validation Zones could be extended to support partial fulfillment.

Reviewer 4: Other than minUTXO concerns, couldn't partial fulfilment be handled by an extra output, and even an extra transaction to merge those outputs?

Reviewer 3: I'm not entirely sure what you mean. Where would the value for the extra output come from? I'm imagining a third-party aggregator trying to settle a big limit order intent against a bunch of smaller intents. An example would be a limit order looking to buy 10,000 ADA being settled against a bunch of smaller limit orders that are only offering 3,000 ADA in total. If a third-party aggregator wanted to batch these orders together using Validation Zones, the aggregator would need to front the difference of 7,000 ADA so that the big limit order intent can actually be used. This is impractical for all, but the most wealthy third-party aggregators. This isn't even an aggregator anymore since fronting the ADA actually makes it a market maker.

Reviewer 4: Limit orders wouldn't be the ones taking advantage of this, because they're long lived and so can afford the contention; they would be implemented as a UTXO on chain that ensures the user is paid their limit price.

It's market orders that would utilize this; the user would construct a transaction in which the token they are selling doesn't appear in the outputs, but the minimum token they're ok receiving is paid to their address.

An off-chain process then matches them against limit orders, other market orders, or AMM liquidity pools in a validation zone; the minimum output is satisfied by whatever liquidity source, because the user doesn't care where it comes from, and if the tx fails because of contention, they can retry and the users signed tx is still valid.

The limit order contract allows partial matches, enforcing that any surplus is paid back into the contract. Since it's sitting there long term, it can afford the contention: it's other people that are spending it, and it doesn't need to provide a signature.

Any surplus over the market orders minimum can be sent to the user in a separate output, barring any minutxo issues, hence my comment above.

Reviewer 3: > Limit orders wouldn't be the ones taking advantage of this, because they're long lived ...

Not all limit orders are long lived. Consider day-traders who are waiting to see if their prices get hit. They can update their limit orders several times within an hour. It would be better if this activity happened off-chain.

> An off-chain process then matches them against limit orders, other market orders, or AMM liquidity pools in a validation zone ...

Perhaps I am misunderstanding something, but AFAIU this is not practical because you can only match against outputs whose values match the off-chain orders *exactly*. This is the double coincidence of wants problem. What are the odds that you can find a limit order or other market order that exactly matches your off-chain order? I believe you even [described](https://github.com/cardano-foundation/CIPs/pull/780#issuecomment-1992362771) in the other PR's discussion that AMMs would first need to break off smaller outputs that then can exactly match the off-chain orders. No DApps currently support "breaking off" outputs, and I am against adding this feature since it involves dramatically increasing the DApp's complexity and attack surface just to make up for a shortfall in the Validation Zone design.

> The limit order contract allows partial matches, enforcing that any surplus is paid back into the contract.

AFAIU partially matching on-chain limit orders against off-chain orders is only possible if the limit order contract supports breaking up the limit order into the fulling output(s) and change output when it needs to be matched against off-chain orders. As I said, I'm against DApps adding support for this.

Reviewer 4: Ok, well, feel free to design things however you want 😅

I was just pushing back agains the claim that the minimum viable design *must* include partial fulfillment. I'm asserting that this design is substantially useful already, and thus serves as a suitable minimum viable design.

Author: It is true that this design is intended for the case where the initial submitter "needs a certain amount of currency X, and is not ok with getting a lesser amount". The original purpose of this project was paying transaction fees in non-Ada tokens. 

I think it might be possible to support more true DEX-like function, but it would require some extra off-chain communication and complexity. The idea is that users must submit requests alongside the requested amounts, to show that they can cover any funds in the requests that were not exchanged for their desired currency. Then they have to spend some of those funds to cover their own un-exchanged requests. Consider the following change to the DEX aggregator usecase :

1. Parties A1 ... An create transactions T1 ... Tn with requests and offers (unlocked outputs) representing their desired trade, as in the previous example.
              - *here, users actually include the amount they request in the inputs AND the outputs AND the requests* 
              - e.g. a request with 3 PolinaCoins must be in a transaction with 3 inputs containing 1 PolinaCoin and 1 output containing 3 PolinaCoins
              
2. Parties A1 ... An send T1 ... Tn to B

3. Party B creates two transactions:
    - T(n+1) has fulfilling all the requests in T1 ... Tn, and spends Party B’s liquidity UTXO `u` in order to do so, creating a new entry `u'` with the remainder. It also requires T(n+2)
             - *at this point, there might still be some tokens missing from the zone, if not all requests have been fully covered by zone transaction, so there are some requests in T(n+1) still. The exchange must keep track of which user's requests are being only partially fulfilled* 
    - T(n+2) spends the offers from T1 ... Tn as well as `u'`, creating a new increased liquidity output `u''`
    
4. B sends the unfulfilled requests (in the form of a UTxO set) to the corresponding users , and gets them to sign transactions T1' T2' which (i) spend the missing requests that remain after the aggregation (and `requireTxs` the transactions in (ii)), and (ii) spend the outputs containing the requested amounts (e.g. 3PolinaCoin output above) to cover the missing requested amounts

5. All parties check and sign the transactions from B that "refund" the unexchanged funds, send them back to B

6. Party B submits a validation zone consisting of T1 ... T(n+2) ; T'1 ; T'2

This is not fully developed usecase yet, but the point is that it uses this infrastructure to to do something that works like partially fulfilling requests.

Reviewer 3: > 4. B sends the unfulfilled requests (in the form of a UTxO set) to the corresponding users , and gets them to sign transactions T1' T2' ...
> 5. All parties check and sign the transactions from B that "refund" the unexchanged funds, send them back to B

IIUC these step introduce a user coordination problem that I think makes this impractical at scale, and will actually create perverse incentives for the aggregator. 

Consider if Alice, Bob, and Charlie all submit off-chain orders as you describe and an aggregator batches them together using Validation Zones. In this batch, all user orders are only partially fulfilled. The aggregator must send the required refund transactions to each user, and wait for them to sign and return the transactions before it can submit the batch.

What if Alice and Bob quickly sign their required refund transactions, but Charlie is unavailable? This would likely mean UTxOs being spent by the batch will be taken by other aggregators/users and this batch transaction is now invalid. Alice and Bob are therefore hurt by Charlie's slow response. This problem will likely happen frequently, especially if this is a global marketplace where all parties are strangers and live in separate time zones. In practice, this means Alice, Bob, and Charlie will need to always be online and quickly sign all refund transactions to maximize the chance of at least one going through. Most of these transactions will likely fail due to the coordination problem. This would be a terrible user experience, and is probably bad enough to deter most people from using this system in the first place.

This is also bad from the aggregator perspective since the aggregator likely only gets paid if the batched transactions are successfully submitted. The more partial users the aggregator tries to batch together, the more the coordination problem will be an issue. To minimize this risk, aggregators will either batch the minimum number of partial users required, or simply avoid any partial filling of orders to avoid the coordination problem entirely.

Author: Those are all fair points.

What it seems like you really need for correctly handling the usecase you propose is something like a "ledger primitive" for a DEX contract, alongside multi-sig and Plutus. That is, a language wherein the only thing you can express is "the tokens in this output are for anyone to take so long as they send some tokens at the specified rate r to my address a", or something like that. All the matching would be done off-chain, eg. by an aggregator. If designed as an extension of the Validation Zones proposal, this would have the added benefit (in addition to cheaper script costs) that transactions with offers do not need to depend on each other or be in particular sequence, so long as all the money in the offers adds up. A transaction that "closes out the zone by paying out the un-exchanged funds" is still needed, but can be made by anyone (no specific signatures required). 

The purposes of these two designs are different enough that they would constitute two different kinds of intents - one where the entire expected payout from an offer can be viewed as funds available for use inside the transaction, and one where the expected payout is not available inside the transaction. It may or may not be possible to generalize these two - as well as all future types of intents - as some unified feature, but this is far beyond the current research being done. Adding different intents as separate features using Validation Zones, however, is much more accessible, especially if we already have evidence that this validation zones usecase (babel fees) works.

Author: The point I am making here is that we consider this CIP as extensible for other kinds of intents, and we would like to asses the response to the Validation Zones design specifically for the case of Babel fees before we introduce more kinds of intents are essentially non-overlapping. 

Would you consider satisfactory a detailed "next steps" section where we discuss a possible design to implement what you are requesting (this would likely not include additional code changes or attack vector analysis, only a specification)?

Reviewer 3: I think I actually came up with an approach to partial matching that doesn't require any changes to the Validation Zone design and also doesn't require "breaking off" pieces of on-chain liquidity sources. AFAICT all of the approaches described so far have been from the perspective of matching each off-chain intent against an on-chain liquidity source. However, it may be more practical to match off-chain intents against each other and *only match the value difference against an on-chain liquidity source*. This approach has the benefit of being able to "on-load" partial fulfillment to the on-chain liquidity sources without having to break up the on-chain liquidity UTxOs.

For a simple example, imagine if Alice has 10 ADA and wants to convert it to 9 DJED. She already has the 10 ADA sitting in a UTxO so she just needs to create the intent transaction, sign it, and send it to an aggregator. This UTxO with 10 ADA will be the fulfilling output for her intent transaction.

Meanwhile, Bob has 8 DJED and wants to convert it to 9 ADA. Like Alice, he already has the 8 DJED in a UTxO so he can just create the intent transaction, sign it, and send it to an aggregator. This UTxO with 8 DJED will be the fulfilling output for his intent transaction.

The aggregator sees the two intent transactions, but can't match them because the values don't match exactly: there is a net value difference of +1 ADA (10 ADA input - 9 ADA output) and -1 DJED (8 DJED input - 9 DJED output). This is the double coincidence of wants problem, but it is easily solvable by using an on-chain liquidity source to convert the extra 1 ADA to 1 DJED. On-chain liquidity sources can be partially fulfilled so it would be easy to find a liquidity source to use.

So the aggregator matches the two intents against each other and matches the value difference against the on-chain liquidity source to produce a valid transaction. The required fulfilling outputs for the intent transactions *already exist* so there is no need for the aggregator to try breaking off pieces of the on-chain liquidity UTxO.

I think this approach works especially well for on-chain liquidity sources that can be composed. For example, if Alice wanted to convert ADA -> WMT and Bob wanted to convert AGIX -> DJED, these two off-chain intents can't be directly matched against each other. However, if there is a liquidity source for WMT -> USDC and another for USDC -> AGIX, the off-chain intents can be indirectly matched together as long as the on-chain liquidity sources can be composed to effectively create a WMT -> AGIX conversion. So this transaction would have 2 off-chain intents + 2 on-chain liquidity sources. It is even possible to batch different pairings of intents+sources as long as the liquidity sources can be composed in the same transaction.

From an incentives perspective, aggregators will actually want to do this, even without charging a fee, because the possible compositions enable finding huge arbitrage opportunities. It is possible for no collateral to ever be at risk with this approach (it depends on the liquidity source logic). I explained these incentives in-depth in my cardano-swaps DEX [readme](https://github.com/fallen-icarus/cardano-swaps?tab=readme-ov-file#naturally-incentivized-liquidity) which uses the same arbitrage possibilities to incentivize aggregators.

This approach also enables users to submit limit orders off-chain, and allows them to update their asking prices off-chain as well (the day-trading use case). The user would need to trust the aggregator to use the newest intent transaction, but all of the cancellation methods mentioned in the CIP would still work with this method so users can revoke permission at any time.

@polinavino Do you have any thoughts on this? Did I misunderstand something that actually makes the above impossible?

Author: Let me try to understand this better - it sounds to me that the following assumptions would be required for what you are suggesting :

- all transactions involved in an exchange of this sort (i.e. contained in the zone) have to be exchanging only the same two types of assets, and they have the same exchange rate between the two. In your example, the exchange rates are different, so I don't quite see why this is ok - but it would work for Alice offering 10DJED for 5Ada, and Bob offering 4Ada for 8DJED.

- the liquidity pools basically absorbs any left over un-exchanged assets by covering any remaining requested tokens and appropriating the extras. In the case I described, DJED pool would be getting 2 DJED and Ada pool would be losing 1Ada

Does this sound right at all?

Reviewer 3: > all transactions involved in an exchange of this sort (i.e. contained in the zone) have to be exchanging only the same two types of assets, and they have the same exchange rate between the two. 

No, both the types of assets and the exchange rates are ultimately irrelevant. AFAIU the only requirements for validation zones are:
- All input "holes" must have a corresponding fulfilling output with the proper value that comes from the proper address.
- The total value of the validation zone must be balanced.

In my simple example, Alice had a UTxO with 10 ADA and created an intent transaction with an input "hole" for 10 ADA from her address, and an output to her address with 9 DJED. To actually execute Alice's intent, the aggregator just needs to specify Alice's UTxO with 10 ADA as the fulfilling output for her intent transaction. The same applies to Bob's intent transaction.

So the validation zone containing both Alice's and Bob's intent transactions has all input "holes" satisfied, but is still unbalanced overall. To balance it, just add another normal input. You don't need to use another intent transaction to finish balancing the zone; a normal transaction, without any input holes, would work. So the validation zone consists of:
- 2 intent transactions and 1 normal transaction
- Alice's 10 ADA UTxO and Bob's 8 DJED UTxO are specified as fulfilling outputs for their respective intent transactions.

All input "holes" are filled and the value of the overall validation zone is balanced.

> - the liquidity pools basically absorbs any left over un-exchanged assets by covering any remaining requested tokens and appropriating the extras. In the case I described, DJED pool would be getting 2 DJED and Ada pool would be losing 1Ada

This part is correct.

Author: ok, i think i see - let me try again - 

There is actually no change to the DEX aggregator example (below), we just **do not make the assumption that the total liquidity in B's pool stays constant**, and instead assume that B participates in the exchange (exchanging some of their own assets) in the case the requests and offers of T1... Tn do not balance out perfectly.

1. Parties A1 ... An create transactions T1 ... Tn with requests and offers (unlocked outputs) representing their desired trade, as in the previous example.
2. Parties A1 ... An send T1 ... Tn to B
3. Party B creates two transactions:
    - T(n+1) has fulfilling all the requests in T1 ... Tn, and spends Party B’s liquidity UTXO `u` in order to do so, creating a new entry `u'` with the remainder. It also requires T(n+2)
    - T(n+2) spends the offers from T1 ... Tn, placing it into liquidity output `u''`
4. Party B submits a validation zone consisting of T1 ... T(n+2)
    - Possibly followed by a transaction aggregating liquidity in `u'` and `u''`

Reviewer 3: > we just do not make the assumption that the total liquidity in B's pool stays constant, and instead assume that B participates in the exchange (exchanging some of their own assets) in the case the requests and offers of T1... Tn do not balance out perfectly.

This part is correct.

> 3. Party B creates two transactions:
    - T(n+1) has fulfilling all the requests in T1 ... Tn, and spends Party B’s liquidity UTXO `u` in order to do so, creating a new entry `u'` with the remainder. It also requires T(n+2)
    - T(n+2) spends the offers from T1 ... Tn, placing it into liquidity output `u''`

This part is incorrect. Only one transaction is required from Party B. T(n+1) uses `u` to finish balancing T1 ... Tn and produces `u'` as a result; this is where the absorption happens. Going from `u` to `u'` required depositing the un-exchanged assets and withdrawing the required assets to finish balancing T1 ... Tn. It stops there. There is no T(n+2) and no `u''`. The validation zone consists of T1 ... Tn and T(n+1). If liquidity sources can be composed, T(n+1) can contain multiple sources that each absorb different parts of the left over un-exchanged assets.

>  T(n+1) has fulfilling all the requests in T1 ... Tn, and spends Party B’s liquidity UTXO `u` in order to do so, creating a new entry `u'` with the remainder. 

Specifically, this part is wrong. The fulfilling UTxOs already exist so there is no need to break off chunks from `u` to fulfill the requests in T1 ... Tn.

Reviewer 3: > The reason there must be two transactions is that it is not possible for a transaction to spend both a regular output and a request from any other transaction.

Does this mean you cannot use a "pre-existing" UTxO to fulfill a request? You must first create the UTxOs in a transaction within the zone for it to be used as a fulfilling UTxO?

For example, is this not possible:
1. Parties A1 ... An create transactions T1 ... Tn with requests and offers (unlocked outputs) representing their desired trade, as in the previous example. The UTxOs meant to fulfill the requests already exist at each parties' address.
2. Parties A1 ... An send T1 ... Tn to B
3. Party B creates two transactions:
    - T(n+1) fulfills all of the request by spending the required UTxOs (that already exist) from Parties A1 ... An.
    - T(n+2) spends liquidity source `u` to finish balancing the zone. This results in `u'` being created.
4. Party B submits the zone consisting of T1 ... T(n+2).

T(n+1) matches the pre-existing fulfilling outputs to the requests for them to be used in T(n+2).

Author: > Does this mean you cannot use a "pre-existing" UTxO to fulfill a request? You must first create the UTxOs in a transaction within the zone for it to be used as a fulfilling UTxO?

It does not mean that, no. It means that if a transaction `tx1` fulfills a request `r` made by transaction `tx2`, it *has* to use a UTxO that is *not* created by `tx2` or any transaction on which `tx2` depends, so that the liquidity used to fulfill `r` could not possibly come from "thin air", i.e. splitting `0` assets into `r` and `-r` assets, and then re-combining these into `0`. 

> T(n+1) fulfills all of the request by spending the required UTxOs (that already exist) from Parties A1 ... An.

This approach requires the "UTxOs (that already exist) from Parties A1 ... An" to be locked by contracts such that they can only be spent if requests are also being fulfilled in the same transaction, which kind of defeats the purpose of using this feature, since the trade is basically done with a contract at that point (which requires more coordination).

In the T(n+1) and T(n+2) i described above, all offers made by A1...An are unlocked. The idea is that even though they are unlocked, the transactions T1...Tn cannot be applied without some other transaction(s) covering the missing value in the requests (which is what T(n+1) does, and also requires T(n+2) to be present), and then T(n+2) collects all the unlocked outputs. No scripts required. Although two transactions are needed, we anticipate that the difference in transaction size between somehow allowing 1 transaction to both fulfill requests and collect offers, and the 2-transaction design we present would likely be O(1), since it is sort of like splitting one transaction into 2, but both parts would be signed separately.

Reviewer 3: @polinavino I appreciate you continuing to answer my questions regarding this CIP! It seems I've been misunderstanding how intents are ultimately modeled in Validation Zones. The more I learn about Validation Zones, the more its design seems to diverge from the way I've always thought Cardano intents would work.

> ### Open (atomic) swaps
>
> A wants to swap 10 Ada for 5 tokens `myT`.
>
> 1. Party `A` creates a transaction `tx1` with
>    - an unlocked request `0` with 5 `myT` tokens,
>    - some inputs totalling 10 Ada
>    - an output `0` which contains 5 `myT` tokens, and
>    - an unlocked output `1` containing 10 Ada
> 
> -- *reordered to group inputs with outputs*

Validation Zones model intents as *completed transactions*, albeit with some holes to fill (the CIP calls these "underspecified transactions"). The above transaction (`tx1`) is fully balanced and specifies *exactly* what each input/output should look like:
- only one extra input is allowed and it must contain 5 `myT`
- only one extra output is allowed and it must contain 10 Ada

But this is not what an intent *actually* is. An intent is just a *piece* of a transaction. This is perfectly exemplified by the above descriptor: `A wants to swap 10 Ada for 5 tokens myT`. This doesn't say anything about the other pieces of the transaction. In the atomic swap above, the intent is more accurately modeled by just:
> - some inputs totalling 10 Ada
> - an output `0` which contains  5 `myT` tokens

The other unlocked input (aka request) and unlocked output in `tx1` do nothing but restrict how this intent can be used:
- You cannot combine intents into a single transaction since they are already complete transactions. Instead, Validation Zones enables batching transactions but this adds cost to the aggregator since each transaction has a minimum 0.15 ADA fee. The fact that all Validation Zone intents are separate transactions means this cost can quickly add up.
- You cannot perform other actions in the intent transaction, even if the intent owner is actually okay with it. If Alice just wants the above atomic swap, she doesn't care if the aggregator also satisfies someone else's intent in the same transaction. The aggregator could also do a swap of their own and Alice still wouldn't care. Forcing each intent to be an isolated transaction seems like a waste of space and money.
- The intent owner has to *guess* what the missing inputs and outputs would look like and place these holes in the transaction. If they guess wrong, this places enormous friction (aka, increased cost) on how an aggregator can actually fulfill this intent. Specifically, the aggregator must contort (ie, break up or merge) other UTxOs to fit the holes set in each intent. The batched transactions in Validation Zones is also meant to address this, but this means the previous bullets are exacerbated since each Validation Zone intent actually requires *multiple* transactions to fulfill it.

I think the eUTxO model actually enables modeling intents more closely to what they are: *pieces of transactions*. Alice can create a piece of a transaction, sign it, and send it to the aggregator. Bob can do the same. The aggregator can take these two transaction pieces and fit them together monoidally (ie, like a puzzle) to create a full transaction. I am making an important assumption here: **a collection of *signed* transaction pieces is just as valid as a single transaction signed by everyone**. I don't see how aggregating the pieces would invalidate the signatures. The fact that each UTxO is atomic seems to guarantee each piece can be considered an *independent event*. This is effectively a multisig where each party only needs to sign the parts of the transaction that are actually relevant to them. It is a *decentralized multisig*.

An empty transaction piece is effectively just a transaction with all fields empty. Users can add pieces they care about and leave the rest blank. So for the above atomic swap, Alice can create a transaction piece with one input from her address for 10 ADA and one output to her address for 5 `myT`; everything else is left empty. Bob can create his intent with one input, 2 outputs, and a validity interval; everything else is left blank. As long as there are no conflicts between the pieces, these two pieces can be fit together to create a transaction with 2 inputs, 3 outputs, and the validity interval from Bob's intent. The aggregator can even add their own piece to the mix to finish balancing it. If Alice and Bob each signed their respective pieces, why does it matter what else happens in the transaction? Alice's input is still only spendable if she gets her output in the same transaction. 

When the completed transaction is submitted, the ledger would verify the transaction is a valid combination of all of the pieces and all pieces are properly witnessed. The pieces and their signatures would need to be somewhere in the transaction. Since the overall transaction can be derived from the pieces, the CDDL can be redesigned around this idea.

For smart contracts, users can specify mints, observers, staking scripts, etc, in there respective pieces. However, unlike how pubkey signatures only need to sign their respective pieces, the smart contracts would be executed against the aggregated (aka completed) transaction. This means smart contracts can even be used to *disallow* certain combinations of intent pieces.

Cancellation would work the same way as in Validation Zones. Just include a reference input or normal input in your transaction piece. To cancel it, just spend the input. These transaction pieces can even contain input "holes" (ie, unspecified output references for certain inputs). I personally don't see a need for output holes; they seem to have been introduced into Validation Zones just to compensate for its restrictiveness.
 
This approach even seems much cleaner to use for Babel Fees. For example, imagine if Alice wants to buy a sword NFT worth 25 GEMS and the transaction fee is worth about 1.4 GEMS (total of 26.4 GEMS). If Alice has a UTxO with 30 GEMS, she can create an intent piece that only specifies an input from her address for 30 GEMS and an output to her address for 3.6 GEMS + the sword NFT; all other parts are left empty. She would sign this piece and send it to the game devs who can now batch Alice's intent piece with other users who also want to purchase game items. Since these intent pieces can easily be combined, it is possible for the game devs to sell a variety of game NFTs to multiple, unrelated users in a single transaction. Doing the same thing with Validation Zones seems overly complicated and expensive (many transactions are required).

I'm not sure of everything that would need to change to support transaction pieces, but on the surface it seems way more expressive and more cost effective than Validation Zones. It doesn't seem like anyone has considered breaking up transactions and signing the pieces like this. I recognize this is probably a major change, but the expressiveness may make it entirely worth it. Is there some reason why it wouldn't work?

@Quantumplation you and I seem to disagree on everything :smile:. Do you have any criticisms of the intent piece concept?

Reviewer 4: I don't have the spare cycles to digest it all, but I don't have any particular problems with it. But, it's just a different system than what's being proposed here. It feels a bit like playing Starcraft 2 and saying "man, this would be great if it was first person and had guns!" 😅

I think what is actually being proposed is powerful enough (or can be extended to support) all the same use cases; just because it doesn't fit what you were picturing doesn't mean it's a deficient design.

I think the main thing that this one provides that yours doesn't (and I'd have to think it through to digest this fully) is probably related to keeping transactions deterministic, and not completely rewriting all of Plutus 😅

Reviewer 3: > I don't have the spare cycles to digest it all, but I don't have any particular problems with it.

All good. I tend to write a lot :sweat_smile:. As long as you think it is reasonable, that is enough for me.

This CIP mentions the following:

> It may be that the best way to implement intents on Cardano is entirely different from the underspecified transactions and validation zones approach we are taking here. In that case such a design might entirely obsolete this one, making it an expensive and pointless burden to maintain in future.

I just want to make sure all possible paths have been considered to avoid this burden.

> I think the main thing that this one provides that yours doesn't (and I'd have to think it through to digest this fully) is probably related to keeping transactions deterministic, and not completely rewriting all of Plutus 😅

AFAIU eUTxO guarantees the determinism even if broken into pieces, but I totally recognize this is currently just an assumption on my part. Even for ordering of the pieces, the pieces can be applied in the order they appear in the list in the CDDL (which can possibly be configurable).

Author: > You cannot combine intents into a single transaction since they are already complete transactions. Instead, Validation Zones enables batching transactions but this adds cost to the aggregator since each transaction has a minimum 0.15 ADA fee. The fact that all Validation Zone intents are separate transactions means this cost can quickly add up.

This is true. Since zones are limited in size to the size of a single transaction, we should consider charging 0.15 ADA per zone (or perhaps slightly more, but not like, 0.15 PER transaction). I will bring this up with the ledger team. 

> You cannot perform other actions in the intent transaction, even if the intent owner is actually okay with it. If Alice just wants the above atomic swap, she doesn't care if the aggregator also satisfies someone else's intent in the same transaction. The aggregator could also do a swap of their own and Alice still wouldn't care. Forcing each intent to be an isolated transaction seems like a waste of space and money.

I dont really get this. You can do whatever you want inside a zone, except use old-version Plutus scripts. For regular atomic swaps, no aggregator is needed, it is just between two parties. The point of designing Validation Zones in the way we did is to reduce the off-chain communication required between parties that want to perform a transaction. None at all is required - you construct a transaction in which you say what tokens you want, and what tokens you are offering in exchange, and a counterparty (you dont care who it is) builds the fulfillment, submitting the complete zone to the network. This is definitely more streamlined than co-constructing a transaction that everyone needs to sign, or a zone that everyone needs to sign. No need for the use of expensive contracts, either. Moreover, the offer and the acceptance of it is done in the same block. You are right in that it might be more expensive because of the number of transactions, however, we may be able to reduce the per-transaction cost. 

I think, actually, all of what you are suggesting is already possible in the current model using co-constructed transactions and smart contracts? 

> The intent owner has to guess what the missing inputs and outputs would look like and place these holes in the transaction. If they guess wrong, this places enormous friction (aka, increased cost) on how an aggregator can actually fulfill this intent. Specifically, the aggregator must contort (ie, break up or merge) other UTxOs to fit the holes set in each intent. The batched transactions in Validation Zones is also meant to address this, but this means the previous bullets are exacerbated since each Validation Zone intent actually requires multiple transactions to fulfill it.

Non-communicating parties that want to perform some exchanges already have to construct individual transactions to put their offers on-chain. I am not sure if multi-party co-constructing of a single transaction that includes their offers and liquidity from an exchange is current practice, but it sound cumbersome, and would still be a big expensive transaction that takes a lot of off-chain communication. In our design, everyone constructs a single small transaction and sends it off. The exchange then constructs two (big) transactions, spending everyones offers and requests. Exchanging lots of peoples offers will always require a transaction with lots of UTxOs (at least one per user).

Reviewer 4: @fallen-icarus 
> AFAIU eUTxO guarantees the determinism even if broken into pieces,

Not true, almost by definition.

First, the assembler has freedom to choose in which order to assemble the pieces; the order of the transaction outputs, for example, can be changed. Additionally, it's impossible to construct the script context to evaluate transactions, and so whoever provides collateral might be putting their collateral at risk, for example. I'm sure there are half a dozen other subtleties that make this a nontrivial thing to spec out a *dramatic* (almost complete) change to the ledger rules / execution model for plutus scripts 😅

top-level transaction.

**Babel fees.** A Babel fee-type transaction is a specific instance of the first use case. A user creates a sub-transaction `tx`
where the missing assets are necessarily a
quantity of Ada. In particular, it does not pay its own fees, collateral, or cover any new `minUTxOValue`s.
The counterparty creates a top-level transaction which includes `tx` as a sub-transaction and includes
extra Ada which goes towards paying transaction fees, collateral, etc.

**DApp fee and min-UTxO sponsorship.** A DApp may choose to subsidize the cost of its use
(e.g. by paying the cost of ExUnits, paying fees/minUTxO, and providing collateral) for the
purposes of encouraging users to interact with it. Implementing nested transactions
will allow building transaction batches that subsidize
DApp-using sub-transactions in this way.

## Specification

For the Agda specification prototype built on top of the current ledger spec,
see [Agda spec](https://github.com/IntersectMBO/formal-ledger-specifications/compare/polina-nested-txs).

### Transaction structure changes

We add the following new field to `TxBody` :

1. `subTxs :: [SubTx]`
-  sub-transactions in the batch

> 2. We also need implementation of the [CIP-112 - Observe script type](https://github.com/cardano-foundation/CIPs/pull/749) in order to allow sub-transactions to require scripts to be executed at the top level with all sub-transactions in their context. "Observe script type" is called "Guard script type in the Ledger implementation, therefore going forward we will use term "gurads", instead of "observers", while refering to the same concept.

Reviewer 2: ```suggestion
2. We also need implementation of the [CIP-112 | Observe Script Type](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0112/README.md) in order to allow sub-transactions to require scripts to be executed at the top level with all sub-transactions in their context. "Observe script type" is called "Guard script type in the Ledger implementation, therefore going forward we will use term "guards", instead of "observers", while referring to the same concept.
```
Spelling and change link to merged CIP from PR

executing transaction are included in the `TxInfo` constructed for its scripts, however, now they don't necessarily have to be attached to 
that transaction.

5. All inputs of all transactions in a single batch must be contained in the UTxO set before any of the 
batch transactions are applied. This ensures that operation of scripts is not disrupted, for example, by 
temporarily duplicating thread tokens, or falsifying access to assets via flash loans.
In the future, this may be up for reconsideration. 

6. The batch must be balanced (POV holds). The updated `produced` and `consumed` calculations sum up the appropriate 
quantities not for individual transactions, but for the entire batch, including the top-level and its sub-transactions.

7. All transactions (sub- and top-level) may specify a non-zero fee. The total fee summed up across all transactions in 
a batch is required to cover the minimum fees of all transactions. The fees specified in all transactions are always collected. 
Individual transactions in a batch do not need to meet the min-fee requirement.

> 8. The total size of the top-level transaction (including all its sub-transactions) must be less than the `maxTxSize`. 

Reviewer 5: This seems ambiguous - does it mean the total size of the top-level transaction including (naturally) the sub-transaction _hashes_, or recursively the total size including all the sub-transaction sizes too?

If the latter, that would seem a shame (and would prevent a particular use case in Leios which this margin is too small to contain), and it's not clear why this constraint is required.

  ,   2  : coin
  , ? 3  : slot_no
  , ? 4  : certificates
  , ? 5  : withdrawals
  , ? 7  : auxiliary_data_hash
  , ? 8  : slot_no
  , ? 9  : mint
  , ? 11 : script_data_hash
  , ? 13 : nonempty_set<transaction_input>
  , ? 14 : guards
  , ? 15 : network_id
  , ? 16 : transaction_output
  , ? 17 : coin
  , ? 18 : nonempty_set<transaction_input>
  , ? 19 : voting_procedures
  , ? 20 : proposal_procedures
  , ? 21 : coin
  , ? 22 : positive_coin
  , ? 23 : [+ sub_transaction] ;; new field
  }

sub_transaction =
  [sub_transaction_body, transaction_witnesses, auxiliary_data / nil]

;; Missing `fee`, `collateral` and `collateral return output`, when comparing to `transaction_body`
sub_transaction_body =
  {   0  : set<transaction_input>
  ,   1  : [* transaction_output]
  , ? 3  : slot_no
  , ? 4  : certificates
  , ? 5  : withdrawals
  , ? 7  : auxiliary_data_hash
  , ? 8  : slot_no
  , ? 9  : mint
  , ? 11 : script_data_hash
  , ? 14 : guards
  , ? 15 : network_id
  , ? 17 : coin
  , ? 18 : nonempty_set<transaction_input>
  , ? 19 : voting_procedures
  , ? 20 : proposal_procedures
  , ? 21 : coin
  , ? 22 : positive_coin
  , ? 24 : [+ required_top_level_guard] ;; new field
  }

> required_top_level_guard = [credential, plutus_data / null]

Reviewer 2: ```suggestion
required_top_level_guard = [credential, plutus_data / nil]
```

Reviewer 6: Nice catch! Thank you!
I don't have access to the branch that was used for submitting the PR, so we'll have to wait for @polinavino to apply this suggestion.

that contain non-empty `requiredTopLevelGuards`. This means this service must construct batches containing these
transactions and also satisfy the batch-level constraints imposed by these scripts being included in the `guards` field of the top level transaction.

An additional high-level language may be beneficial to specify what the guard scripts
actually require of the batch, as Plutus script constraints may be difficult to work with directly.

## Rationale: how does this CIP achieve its goals?

The primary purpose of this CIP is to enable Babel Fees. Our design allows users to perform arbitrary
swaps without two-way off-chain communication, and with counterparty irrelevance. This simulates
fee payment in arbitrary currency whenever another use is willing
to accept the non-primary asset offer in exchange for paying the transaction fee. This is the core
requirement of Babel fees. In addition, the goal of this design is to lay the groundwork for
supporting more general intents and the possibility of extending to additional ones.
We have done so by adding the following set of features :

> 1. transactions can be batched

Reviewer: This "set of features" already appears more or less exactly in the _Motivation_.  I can see how it would be at home in either section: since it indicates both the feature that are needed _and_ why they've been conceived this way.  But @polinavino (cc @lehins) could you remove the duplicate in any case & ensure it's the latest content there (in case the 2 copies have diverged during editing)?

Reviewer: resolved by https://github.com/cardano-foundation/CIPs/pull/862/commits/6e4e887de4199514fd141e79966f1fe23f16d288

- batch contains one top-level transaction, with multiple possible sub-transactions ;

2. the batch must be balanced, but individual transactions in it do not have to be, so
- transactions in a single batch may be used to pay parts of the fee and satisfy `minUTxOValue` requirement for other sub-transactions;
- unbalanced transactions can be interpreted as swap offers, getting resolved within a complete batch ;

3. script data is shared across all transactions in a single batch
- enables script deduplication ;

4. batch script guards enable sub-transactions to specify properties required of batches in which they may be included
- represents all possible intents that do not require ledger rule relaxation ;

5. the top-level transaction must provide collateral for all scripts in its sub-transactions
- sub-transactions are neither required nor allowed to cover their own collateral ;

> ## Use cases

Reviewer: resolved by https://github.com/cardano-foundation/CIPs/pull/862/commits/6e4e887de4199514fd141e79966f1fe23f16d288

trading tokens it is interested in, and are being traded at a favourable price.
It may or may not require incoming transactions to cover their own fees and/or collateral.
In the case a set of sub-transactions is not fully balanced, and the aggregator is interested
in participating in the exchange offers within the batch, it would have to include its own
liquidity pool as one of the inputs
in the top-level transaction, and output any extra tokens to its updated liquidity pool UTxO.

### Babel fees

Babel fees is a specific subtype of the first usecase where the missing assets are necessarily a
quantity of Ada. Usually, it also implies that no Ada is contained in the inputs of such a
transaction, aside from the minimal amount enforced by `minUTxOValue` requirement, and the missing
Ada goes towards paying transaction fees.

### DApp fee and min-UTxO sponsorship

DApp may run a Babel service which only cares only about transactions interacting with a
certain DApp, discarding all other incoming transactions immediately. This service
may include calculating ExUnits, paying fees/minUTxO for, and providing collateral for such
partially valid transactions.

> ## Pseudo-code Example

Reviewer: Likewise we have no top-level section by this name, so unless moving to the end in an _Appendix_ this would be part of the _Specification_ (OK since the "use cases" should probably be introduced in the _Motivation_ section immediately before the spec).

are :

1. The top-level transaction contains subTxs, which
are lists of transactions. They are processed as individual transactions, alongside the top-level tx. As a result,

- the `TxIn` of each output can be computed for each output partially valid transaction
at the time of construction

- the `TxId` of the transaction for each entry in the UTxO set is necessarily
signed by all the keys required by that transaction

- the input to (and therefore output of) each script is predictable at the time of (partially valid) tx construction,
so that it is possible to compute required `ExUnits` for each of the scripts run by every sub-transaction
at the time of construction

2. Contracts can only view data in the transaction running them, except in the case
that a top-level tx is running a guard script

3. In this design, script outputs (except of guards) can be computed by any Babel service receiving an incoming
partially valid transaction (as well as the author of the transaction), and they do not depend on the top-level transaction.
This makes constructing (and determining whether it is even possible to construct)
a valid top-level transactions more straightforward.

> ### CIP-0118 "Validation Zones-implicit"

Reviewer: ```suggestion
#### CIP-0118 "Validation Zones-implicit"
```
(continued from above)

and Transaction Swaps. The following 
new usecases and functionality are supported by the changes :

1. transactions can be batched
- batch contains one top-level transaction, with multiple possible sub-transactions ;

2. the batch must be balanced, but individual transactions in it do not have to be, so 
- transactions in a single batch may pay each others' fees and minUTxOValue ;
- unbalanced transactions can be interpreted as swap offers, getting resolved within a complete batch ;

3. script data is shared across all transactions in a single batch 
- enables script deduplication ;

4. batch observer scripts enable sub-transactions to specify properties required of batches in which they may be included 
- design inspired by script observers ;

5. the top-level transaction must provide collateral for all scripts in its sub-transactions 
- sub-transactions are not required to cover their own collateral ;

6. a new intent we call "spend-by-output", wherein a sub-transaction may specify outputs it 
- intends to spend, and the top-level transaction specifies the inputs that point to those outputs in the UTxO set ;
- example of possible additional supported intents (see CPS-15)

> potentially useful for LC functionality ;

Author: This feature is removable, and I can edit the CIP in the coming week(s) to do so, after discussing it at the meeting Monday.

must satisfy without violating existing ledger rules. Adding support for intents that *do require ledger rule relaxations*, such 
the ones in (2, 3, 5, 6) above, must be done 
more carefully to ensure secure operation of the ledger program.  

## Specification

For the Agda specification prototype built on top of the current ledger spec, 
see [Agda spec](https://github.com/IntersectMBO/formal-ledger-specifications/compare/polina-nested-txs).

### Missing and extra assets 

For a given transaction body, let `produced` and `consumed` be the produced and consumed 
values, respectively. Then

extra assets/currency `:= { aid ↦ v | aid ↦ v ∈ consumed - produced , v > 0 }`

missing assets/currency `:= { aid ↦ v | aid ↦ v ∈ produced - consumed , v > 0 }`

### Transaction structure changes

We add the following new fields to `TxBody` : 

1. `requireBatchObservers : ℙ ScriptHash`
- set of scripts that must be run by the top-level transaction in the batch (where `TxInfo` includes info of all batch transactions)

2. `spendOuts : Ix ⇀ TxOut`
- (indexed) outputs being spent 

3. `corInputs : ℙ TxIn`

> - inputs corresponding to `spentOuts`

Author: these have to do with the "spend-by-output" feature

4. `subTxIds : ℙ TxId`
-  IDs of sub-transactions in the batch

The following field is added to `Tx`:

1.  `subTxs : List Tx` 
- list of sub-transactions

**Note** In the Agda spec, the transaction body contains `subTxIds : ℙ TxId`. The Haskell implementation aligns 
with this to avoid confusion. A better design will be to include the full sub-transactions 
directly in the body of the top-level transaction, making `Tx` a recursive type. 

> However, the recursive types design is more difficult to represent in Agda. 

Author: I'll think about this some more! (and then we will talk about it at the meeting, I imagine)

> However, the recursive types design is more difficult to represent in Agda. 

Reviewer 7: I'd like to point out that there's no Agda limitation here. You can do recursive records just fine, you just need to decide whether they are inductive or coinductive, in this case you want inductive.

I also agree that even if there was a limitation here, it's not an argument to constrain the design. There are plenty of techniques that we're already using to work around such issues, and in general every program can somehow be represented in Agda, even if it is really ugly. So such an argument really only works for inconsequential design decisions.

**Note** The era which will include this CIP may be combined with the following additional changes :

- `RdmrPtr ⇀ Redeemer` structure inside the body
- datums inside the body
- `requiredObservers` scripts that are checked by a transaction but are not for the purpose of validating a specific action (see 

> [CIP-0112](https://github.com/cardano-foundation/CIPs/pull/749))

Reviewer: cc @colll78
> CIP-112 is definitely a prerequisite for Nested Transactions

### Plutus

#### New version

A new Plutus version, `PlutusV4`, will be required, since the constraints on what it means for 
a transaction to be valid, as well as the tx structure itself, are changed. Transactions 
that are top-level transactions with no sub-transactions are still able to interact 
with V3 scripts. Transactions that are underspecified and/or using any new features 
introduced here are not. 

#### New script purposes

The following two new script purposes are introduced (together with script purpose tags with the same names) : 

1. `SpendOut : Ix → ScriptPurpose`, where the index indicates that a script is being run to validate spending a specific 
output in `spendOuts`.

> 2. `BatchObs : ScriptHash → ScriptPurpose`, where the script has indicates the script that must be run by the top-level 

Reviewer: > give ability to the subtransaction builders to supply some Data. This might be solved by https://github.com/cardano-foundation/CIPs/pull/735

cc @colll78 - and editors (cc @Ryun1 @Crypto2099 @perturbing) to consider this relationship in that complicated review process

> 2. `BatchObs : ScriptHash → ScriptPurpose`, where the script has indicates the script that must be run by the top-level 

Author: > Hash is probably not gonna be enough, because multiple sub transactions can specify the same script hash. So, instead I think we'll need to use a tuple of indices (Int: index of the subtransaction, Int: index into the list of specified observers).

I was thinking that since these batch observers are constraints on a batch, a script together with a `BatchObs` tag would be enough to specify the script purpose. I imagine each batch observer is executed exactly once. Extending to a per-sub-transaction batch observer execution (which includes allowing each sub-transaction to provide a distinct redeemer) is definitely doable, but i worry that it would make it that much more complicated for a batcher to satisfy the constraints of the batch observer.

> 2. `BatchObs : ScriptHash → ScriptPurpose`, where the script has indicates the script that must be run by the top-level 

Reviewer 6: > that it can be run just once by the top-level tx, and give a different outcome depending on the set of sub-txs

You can't execute a script once and have different outcome. In order to have a chance of different outcome a script must be executed multiple times with different arguments.

> 2. `BatchObs : ScriptHash → ScriptPurpose`, where the script has indicates the script that must be run by the top-level 

Reviewer 7: I think this is just an issue of wording. It wouldn't be multiple executions, it would just be the single execution of the script iterating over all relevant subtransactions.

I have a different comment here, which is that I don't think `BatchObs` should have any arguments at all. As specified, a script run with that purpose will just be passed its own hash which isn't terribly useful.

> 2. `BatchObs : ScriptHash → ScriptPurpose`, where the script has indicates the script that must be run by the top-level 

Reviewer 6: It is not an issue of wording. I believe it as in issue of confusing two different features. We have a [CIP-112 Observer type](https://github.com/cardano-foundation/CIPs/pull/749) that specifically designed to deal with top level scripts that run once for the whole transaction. Which would match the semantics of what you are talking about: "single execution"

In this case we need for sub-transaction to be able to specify arguments to a script that will run at a top level. Which means if two separate sub-transactions specify the same script but two different arguments, you cannot run the script "once", you need to run it as many times as it appears in all of the subtransactions with different arguments. This is a crucial feature for nested transactions, because there would be no other way for sub transaction builder to run and influence the outcome of script at the top level that is unique to that sub-transaction.

> 2. `BatchObs : ScriptHash → ScriptPurpose`, where the script has indicates the script that must be run by the top-level 

Reviewer 6: It is akin to saying "these two certificates are locked by the same script, we should be able to run that script only once". Which, naturally is not the case, because they will have different arguments and different execution units, hence will be executed multiple times. Same concept will be applied to required observers for sub-transactions.

> 2. `BatchObs : ScriptHash → ScriptPurpose`, where the script has indicates the script that must be run by the top-level 

Reviewer 7: So there's a design space here, and the question is how to deal with multiple subtransactions requiring the same observer. The observer could either run once in total or once per subtransaction that requires it, both of these are valid options. So the question becomes which design is preferred here.

One reason why it seems more natural to me to have a single execution of the script is that subtransactions cannot provide a redeemer in the current design, since it wouldn't be signed. So all redeemers would have to be provided at the top level transaction, but then you could just simplify this to a single execution anyway. That has the benefit of better performance which is nice.

So, the question becomes: is there a use case for multiple executions of a batch observer that makes it worth the extra effort of adding signed redeemers? Since we're probably adding regular observers as well, I don't think consistency with other script purposes is really an argument here anymore.

I don't know which option @polinavino originally had in mind here, but with the way CIP is currently designed single execution seems more natural.

> 2. `BatchObs : ScriptHash → ScriptPurpose`, where the script has indicates the script that must be run by the top-level 

Reviewer 6: > One reason why it seems more natural to me to have a single execution of the script is that subtransactions cannot provide a redeemer in the current design, since it wouldn't be signed. So all redeemers would have to be provided at the top level transaction, but then you could just simplify this to a single execution anyway.

This is actually incorrect. Sub transaction will be supplying redeemers and will be signing them, all like with a regular transaction, except for the observers that they require to be executed at the top level since they can't possibly know the amount of execution units that will be required for running such script.

> So, the question becomes: is there a use case for multiple executions of a batch observer that makes it worth the extra effort of adding signed redeemers?

The answer is: it is a requirement. Not for all cases, but without ability of a sub-transaction builder to supply an argument to a script it has no way of affecting the outcome of execution of such script, unless the author hard codes the values, in which case such script cannot be shared with any other sub-transaction anyways.

I have talked to Ziyang about this and he is 100% agreement with me that without ability to provide arguments to a script this feature is practically useless. So, having nice performance isn't very useful if a feature is useless.

In any case. I believe I have a solution that would allow the user to choose which approach suites their use case best. It requires CIP-112, which we were planning to implement anyways for the next era.

We add to the sub transaction body a list of `ReqObservers`:
```haskell
data ReqObserver = ReqObserver
    { reqObserver :: !(Credential 'Observer)
    -- ^ Key hash, or script hash of a plutus or native script
    , reqScriptData :: !(Maybe Data)
    -- ^ In case of a native script or a key hash it is a failure to supply this argument. In case of
    -- a plutus script this is an optional argument. If data is supplied for a plutus script then it
    -- will be executed as `SubTxObserverPurpose` otherwise it will have to be included 
    -- in the `reqObservers` of top level transaction, by the top level Tx builder
    }
```

In other words if a user specifies an argument to a script then it will be a unique execution regardless if other sub-transactions use the same script, but if the argument is not supplied then such script will have to be placed into requiredObservers at the top level, which will lead to de-duplication of execution if other sub-transaction use the same script also without supplying arguments.

> 2. `BatchObs : ScriptHash → ScriptPurpose`, where the script has indicates the script that must be run by the top-level 

Reviewer 6: See my comment here for proof-of-concept definition of sub transaction type and needed changes for top level transaction: https://github.com/IntersectMBO/cardano-ledger/issues/5045#issuecomment-2882021332

> 2. `BatchObs : ScriptHash → ScriptPurpose`, where the script has indicates the script that must be run by the top-level 

Reviewer 7: > This is actually incorrect. Sub transaction will be supplying redeemers and will be signing them, all like with a regular transaction, except for the observers that they require to be executed at the top level since they can't possibly know the amount of execution units that will be required for running such script.

Note that I was talking about the "current design" which didn't have a way to pass data from subtransactions, this has just been added.

This also adds a bunch of extra complexity, since assembling the script context now needs to fetch data from multiple sources, and the ability to provide those needs to be part of the CDDL. So now you need to fetch execution units from the top level, and the redeemer from the subtransaction.

>> So, the question becomes: is there a use case for multiple executions of a batch observer that makes it worth the extra effort of adding signed redeemers?
>
> The answer is: it is a requirement. Not for all cases, but without ability of a sub-transaction builder to supply an argument to a script it has no way of affecting the outcome of execution of such script, unless the author hard codes the values, in which case such script cannot be shared with any other sub-transaction anyways.

Claiming it's a requirement doesn't provide a use case. People are already complaining a lot that decisions are made without providing reasoning, so can we please actually provide an example where this is useful?

Also, it seems like you haven't sufficiently explored the design space here. The question of 'can a subtransaction pass data to its required observer?' is independent from the question of whether multiple executions are required.

In fact, with the data structure `ReqObserver` you've provided, if we properly include it in the `TxInfo` this issue is already solved even with a single execution. This data would just be visible from a singular top-level execution, so the script can just iterate over subtransactions and fetch the relevant data from `ReqObserver`. This seems to have much better properties in my eyes:
- none of the extra complexity in assembling the script context, and
- better performance.

So again, is there a use case that actually requires multiple executions? I can't think of any.

> 2. `BatchObs : ScriptHash → ScriptPurpose`, where the script has indicates the script that must be run by the top-level 

Author: > This data would just be visible from a singular top-level execution, so the script can just iterate over subtransactions and fetch the relevant data from ReqObserver

My initial intuition was this also. However, you cant actually condition on the `TxId` in the sub-tx that included the script in the requirement, because you only get that after constructing the transaction in full. Do we care about this? I have no idea

> 'can a subtransaction pass data to its required observer?' 

I initially thought that the scripts can be written in a way thats sort of pre-applied to the data they want. Alexey said this is not a good programming practice and greatly reduces possibility of script sharing. I also do not know if we care about this.

On both counts, though, I the increase in complexity isnt too bad and is quite confined (see Agda spec)

> 2. `BatchObs : ScriptHash → ScriptPurpose`, where the script has indicates the script that must be run by the top-level 

Reviewer 6: @WhatisRT

> Note that I was talking about the "current design" which didn't have a way to pass data from subtransactions, this has just been added.

That is not my fault that CIP was stale for a while, I have plenty of other responsibilities to worry about. I've made this particular point long time ago and it can be found in many of my comments on this PR.

> In fact, with the data structure ReqObserver you've provided, if we properly include it in the TxInfo this issue is already solved even with a single execution. This data would just be visible from a singular top-level execution, so the script can just iterate over subtransactions and fetch the relevant data from ReqObserver.

It is not solved and in fact what you are suggesting is impossible! It is impossible because the script cannot know if the argument was intended for that script or some other script. The fact that it is visible, does not make it usable, because script cannot possibly know what is "relevant" what is not.

Let's look at a concrete example. Let's say we have two sub-transactions that use the same `script1` with `scriptHash1`, but with different arguments and we have some other scripts with their own arguments.

```haskell
Tx
  { inputs = ...
  , subTransactions = 
    [ SubTx
      { inputs = ...
      , reqTopLevelObservers = 
          [ ReqObserver scriptHash0 data0
          , ReqObserver scriptHash1 data1
          , ReqObserver scriptHash2 data2
          ]
      }
    , SubTx
      { inputs = ...
      , reqTopLevelObservers = 
          [ ...
          , ReqObserver scriptHash1 data3
          , ...
          ]
      }
    ]
  }
```

What I am saying is that the `script1` that corresponds to `scriptHash1` cannot be executed only once, because it needs to be executed once with `data1` as an argument and once more with `data3` as an argument. 

With your suggestion `script1` that corresponds to `scriptHash1` should be able to find both `data1` and `data3` that where intended for it through iteration over all sub-transactions. That suggestion implies that `script1` knows its own `scriptHash1`, because that is the only way it could possibly know they were intended for `script1`. See the problem? Without specifying the redeemer for each of those multiple executions the `script1` that hashes to `scriptHash1` cannot know those arguments are intended for the `scritpHash1` simply through looking at those sub-transactions transaction, because that would imply the script knows its own hash! It also cannot treat all `reqTopLevelObservers` equally since there will be other scripts that will have their own data, like `data2` that is intended for `script2` with `scriptHash2` and will probably be uninterpretable by `script1`

> People are already complaining a lot that decisions are made without providing reasoning ...

Right now you are the only one that is complaining. I've had a few Ledger working group meetings that focused on Nested Transactions. I had discussion both with the community members and plutus devs. There have been plethora of discussions on multiple related CIP PRs about this feature. So far everyone agreed that it is a necessary feature.

It has been agreed that as a sub-transaction builder there must be a method to specify a script that can be executed with the whole transaction in its context. However, without ability to supply arguments to such script, sub-transaction builder has no control over the outcome of execution of such top level script! That is my argument.

In other words, someone gives you a function that you can use (eg. some script that has already been published on chain in a form of an reference script in UTxO), but you have no means of securely supplying an argument to that function. Of what use that function is to you? This is my reasoning and I believe it is very important. Frankly, I am quite surprised that you don't see it that way.

> ..., so can we please actually provide an example where this is useful?

Here is a simple example that comes to mind. Say there is a DApp that will run successfully only if a sub-transaction that specifies it also supplies a cryptographic signature of some nonce. So, in order to use that script a person that builds a subtransaction need to be able to securely supply the public key and a signature to that script for verification. How would you do this without the design that I am proposing? Where in the subtransaction would I put this data, in such a way that it cannot tempered with and can be discovered by that script? It would be unreasonable and insecure for the script to try and look at all redeemers or all data arguments supplied in all subtransactions. Not to mention that it would also be very expensive.

> This also adds a bunch of extra complexity, since assembling the script context now needs to fetch data from multiple sources, and the ability to provide those needs to be part of the CDDL. So now you need to fetch execution units from the top level, and the redeemer from the subtransaction.

Reading this comment makes me realize that you don't fully understand the design that we've converged on, which is fine and I'll be happy to help you understand it better. If you are really interested in learning about it and you'd like to participate more and provide your feedback, I'll be happy to discuss it with you in a live meeting. Please come and join ledger working group next Monday https://lu.ma/1ucg1qyr and I'll be happy to go over any concerns that you might have. We are at a point that I'd still be very happy to find a different or modified approach that solves the same problems but in a better, cheaper and faster way. So far, that is the best that we've got.

> 2. `BatchObs : ScriptHash → ScriptPurpose`, where the script has indicates the script that must be run by the top-level 

Reviewer 7: @polinavino 

> I initially thought that the scripts can be written in a way thats sort of pre-applied to the data they want. Alexey said this is not a good programming practice and greatly reduces possibility of script sharing. I also do not know if we care about this.

I agree that forcing scripts to include the data is probably not the best idea, though it might be fine if there's a really good design reason (which I don't think there is in this case).

> On both counts, though, the increase in complexity isnt too bad and is quite confined (see Agda spec)

The particular complexity I'm worried about is the logic constructing redeemers, which would need to fetch the `ExUnits` from the top level and the `Data` from the sub transaction. That logic does not appear to be in the spec you've written so far. With a single execution model, there is no such logic required.

@lehins 

> That is not my fault that CIP was stale for a while, I have plenty of other responsibilities to worry about. I've made this particular point long time ago and it can be found in many of my comments on this PR.

Yeah, but then maybe don't claim what I'm saying is incorrect especially since this is not the final design. Especially since the design you referred to isn't even sketched out in any of the comments you made about it.

> It is not solved and in fact what you are suggesting is impossible!

So the problem that you seem to have here is that the script doesn't know its own hash to correctly look up the relevant data. Here are two ways of doing 'the impossible' (as you phrased it):
1. We actually know the script hash before running the script. We can just pass it into the script context directly.
2. We collect all the relevant data by iterating over subtransactions before the script, turn it into a map `TxId ⇀ Datum` and pass that into the script context.

> Reading this comment makes me realize that you don't fully understand the design that we've converged on, which is fine and I'll be happy to help you understand it better. If you are really interested in learning about it and you'd like to participate more and provide your feedback, I'll be happy to discuss it with you in a live meeting.

I'm not interested in getting an explanation for myself, I want things to be documented. If there's something I don't understand after reading this CIP & the comments I'm sure most people (who have way less detailed knowledge of the Ledger) won't understand it either. The Ledger working group meeting doesn't seem to be recorded or have meeting notes, so I'd rather discuss things here.

> Right now you are the only one that is complaining.

Yes, I'm complaining now and once this feature is live people from the community who have to deal with it will be complaining. It happens every time, so it'd be a good idea to do something about it this time.

> 2. `BatchObs : ScriptHash → ScriptPurpose`, where the script has indicates the script that must be run by the top-level 

Author: > The particular complexity I'm worried about is the logic constructing redeemers, which would need to fetch the ExUnits from the top level and the Data from the sub transaction. That logic does not appear to be in the spec you've written so far. With a single execution model, there is no such logic required.

Yeah, the spec was wrong w.r.t. logic of getting redeemers and the script data. I think it is fixed now.

> 2. `BatchObs : ScriptHash → ScriptPurpose`, where the script has indicates the script that must be run by the top-level 

Reviewer 6: > I'm not interested in getting an explanation for myself, I want things to be documented.

@WhatisRT It does not seem like you only just want a better documentation, from your suggestions it seems like you want to actively participate in the design of the CIP. Which is great, I love hearing your opinion. However, for this to actually work we need to be on the same page about what we have today. So, instead of coming in and throwing a bunch of random suggestions, without understanding the current design, it would have been better if you participated in some of our prior discussions, or at the very least would be willing to learn about it. That was the reason why I invited you to a ledger working group meeting, which is designed precisely for that: communication about what we are working on and getting some feedback on it, so we can see if we should pivot in any way.

> The Ledger working group meeting doesn't seem to be recorded or have meeting notes, so I'd rather discuss things here.

Every meeting so far has been recorded. You can find the recordings on each individual meeting here: https://lu.ma/calendar/manage/cal-Zl1xd0BB5UoRehN
We've discussed Nested Transactions at the latest [meeting #16](https://lu.ma/1ucg1qyr) that happened today as well as a few other prior meetings in the past. Feel free to watch them if you care about learning on how we landed on the design that we have today.

> If there's something I don't understand after reading this CIP & the comments I'm sure most people (who have way less detailed knowledge of the Ledger) won't understand it either. 

You are absolutely correct on that, so for the future I'd recommend you:
* either wait for the CIP to be ready for review, before you start complaining that you don't understand something
* or even better participate in some discussions where we make decisions on how to design a feature, in which case you would not only be able to actively participate in the design of the feature, but you'd also be able to submit suggestions with corrections to the CIP, because at that point you would already have a clear understanding on how the CIP should look like.

> .. then maybe don't claim what I'm saying is incorrect especially since this is not the final design. Especially since the design you referred to isn't even sketched out in any of the comments you made about it.

@WhatisRT I always value your opinion, but I believe it is counterproductive for you to come out of the blue and say things along the lines "why don't you do it in this way instead", when you don't even have a clear picture of the current design. In other words, the comment you made about me making claims about what you are saying is incorrect is exactly what you are doing: you claiming that current design is incorrect without understanding it. You probably do have now an understanding of it, but I really don't like how we got there.

Some of the suggestions you made do have merit, but without discussing the details of design and evaluating pros and cons of different approaches, we cannot make a decision. So, I'll say it again, if you'd like to have some impact on the CIP, then let's organize a meeting where we can have a productive conversation about the design choices  that we can make. Otherwise these design choices will be made without you, which is fine by me, but then, please, do not come back a year later complaining that we did not take your suggestion.

Side note: simplicity of implementation in Ledger should not be the primary motivator behind choosing one design over another, therefore we need to get some feedback from the community whenever we make a choice that impacts DApp developers. Which means we could include alternative approaches in this CIP and get some feedback that way, but before we even do that we need to evaluate them ourselves for viability.

> Yes, I'm complaining ...

That is precisely the thing that I have a problem with in this thread and I would appreciate it if it could be avoided it in the future. I prefer to have a discussion with contributors, whoever that might be, than people jumping in with complaints, especially during the design phase.

transaction to observe the validity of the batch.

`ScriptsNeeded` and `getDatum` are modified accordingly to include scripts for these purposes. `collectPhaseTwoScriptInputs` 
also requires an additional parameter for constructing `TxInfo` to either contain lists of `TxInfo` for sub-transactions or not.

#### TxInfo

The `TxInfo` shown to a PlutusV4 script has an added field :

- `otherInfos : List TxInfo`, which is populated with the `TxInfo` data for all sub-transactions in the batch 
whenever both hold : (1) the transaction for which info is being constructed is top-level, and (2) the script purpose for 
which this info is constructed is `BatchObs`.

### Ledger Transition Changes

The prototype for the changes can be found
[here](https://github.com/IntersectMBO/formal-ledger-specifications/compare/polina-nested-txs)

The transition structure is changed from 

`LEDGERS -> LEDGER -> UTXOW -> UTXO -> UTXOS` 

to 

> `LEDGERS -> LEDGER -> SWAPS -> SWAP -> UTXOW -> UTXO -> UTXOS`

Reviewer 6: It doesn't really matter. This should be removed from the CIP anyways, since that is an implementation details that is not relevant to the stake holders. It is much better to add some examples of actual transactions instead, as Kevin pointed out in his comment

> `LEDGERS -> LEDGER -> SWAPS -> SWAP -> UTXOW -> UTXO -> UTXOS`

Reviewer 6: It doesn't really matter. This should be removed from the CIP anyways, since that is an implementation detail that is not relevant to the stake holders. It is much better to add some examples of actual transactions instead, as Kevin pointed out in his comment

checks are :

1. Fees are correct. `feesOK` is modified by (1) replacing `txfee` with 
the sum total of `txfee` values of all the transactions in the batch.
The per-transaction fee is collected only once for the batch. 

2. No regular inputs in any transaction are in also included as `corInputs` of the top-level tx

3. All `txins` and `corInputs` are contained in the UTxO set to which the top-level tx is being applied 

4. `corInputs` in top-level tx point to the same (unordered) list of outputs in the UTxO set 
as the list given by `spendOuts` 

5. If the batch is not made up of a single transaction with failing scripts, the batch 
must be balanced (POV holds). Both sides of the equation of POV now sum all of what used to be the 
`produced` and `consumed` values for individual transactions. The value spend by `corInputs` 
is added to the `consumed` side.*

6. Transaction size (which includes top-level and all sub-transactions) is below max for a single transaction

7. The sum of the execution units of all transactions in a batch is less than the max for a single transaction 

8. The list of IDs of `subTxs` contains the same IDs as in `subTxIds` 

> 9. There are no repeating transaction IDs in `subTxs`

Author: i'm gonna have to ask @WhatisRT how to make this work (it wasnt straightforward because Agda complained about the associated proofs required to construct a `Map` type that says that there are no repetitions in the domain)

collecting collateral only, so none of their other actions, including moving assets, will be applied,
and therefore will not affect the ledger.  

To construct the transaction list representing the complete batch, the original transaction prepends the 
top-level transaction to its `subTx` list.

The `LEDGER` rule then calls the `SWAPS` rule on the batch list, and the same state as it itself 
has. The result of the `LEDGER` state update is computed by `SWAPS`. 
However, the environment for `SWAPS` is distinct from `LEnv` (see below). 

**Batch Information**  We define a new type, `BatchData`, to allow `LEDGER` to indicate to `SWAPS` via its environment
the batch info required for correctly processing it. The type has constructors :

- `SingularTransaction`, which indicates that the batch contains only one transaction with no sub-txs, but this tx contains new features

- `OldTransaction`, which indicates that the batch contains only one transaction with no new features (old version of Plutus can be run)

- `BatchParent txid batchValid` which indicates that this batch contains a top-level tx with ID `txid`, and it has sub-txs. The conjunction of 

> all `isValid` tags of all transactions in the batch (sub- and top-) is `batchValid`.

Author: these are indicators of how to handle the entire batch (i.e. whether to collect collateral). All other rules for processing transactions in a single batch are almost exactly the same. If we were to forbid non-zero collateral and remove the "spend-by-output", they would be exactly the same

1. `Scripts-Yes` : in the case that `validPath` is true, it removes both the 
`txins` and the `corInputs` from the UTxO set (in addition to all the existing 
changes done previously by `Scripts-Yes`).

2. `Scripts-No` : in the case `validPath` is false, and `isTop` is false.
No changes are made to the state here because collateral is only collected from the top-level transaction.

3. `Scripts-No-TopLevel` : in the case `validPath` is false, and `isTop` is true.
Collateral is collected, as in the previous version of the `Scripts-No` rule. 

### Collateral and Phase-2 Invalid Transactions

Enough collateral must be provided by the top-level transaction to cover the sum of all the 
fees of all batch transactions. The collateral 
can only ever be collected from the top-level transaction.
The sub-txs are not obligated to provide sufficient 
collateral for their own scripts, but they may choose to (this is never checked for sub-transactions). 
A Babel fees service [Babel Service](#babel-service)
may or may not require incoming underspecified transactions to provide their 
own collateral, depending on whether they want to assume the risk of not being 

> compensated for running failing scripts. 

Reviewer 6: I think we should disable `collateralReturn` and `collateralInputs` fields at the CDDL level. It will simplify the rules bit.

If a sub-transaction passes required phase-1 checks (excluding being balanced/POV, and checks related to `spendOuts` and `corInputs`),
it can be posted on-chain as a `SingularTrasaction`, so that it will be processed as top-level, and its collateral will be collected.

### Batch Observers 

Batch observers are scripts that must be run by a top-level transaction, but may be required by its sub-transactions 
as a way to specify batch-level constraints. The have the script purpose with the tag `BatchObs` and the identifier `ScriptHash`.
These are the only scripts that, in their `TxInfo` data, get a non-empty `otherInfos : List TxInfo` field, which is populated 
with the infos of the sub-transactions. This is inspired by, and may be implemented in conjunction with the 
[Observe Script Type CIP](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0112). 

The difference is that for a tx-observer script in the linked CIP is run on the `TxInfo` of the transaction 
that requires it. With batch observers, the scripts are required by a sub-tx, but run with top-level input. 

> For the same top-level transaction, a regular observer and a batch observer in a transaction will get different `TxInfo`. 

Author: To construct TxInfo, you need a `ScriptPurpose`. If the script purpose is `BatchObs sh` , the `subTxInfos` is populated (otherwise it isnt). This script purpose is also allowed to have an associated redeemer. If we change this script purpose to specifying the transaction for which it is the observer (via index Int or via TxId directly), we can have multiple redeemers for the same script hash

> For the same top-level transaction, a regular observer and a batch observer in a transaction will get different `TxInfo`. 

Reviewer 6: > To construct TxInfo, you need a ScriptPurpose.

No you don't.
`TxInfo` is always the same for all scripts that are executed in a transaction. In fact we even memoize it in order to avoid computing it multiple times.

I believe you are thinking of `ScriptContext`, which itself contains `TxInfo`, `ScriptPurpose` and `Data`

> For the same top-level transaction, a regular observer and a batch observer in a transaction will get different `TxInfo`. 

Reviewer 6: That is true, top level transaction will have a different TxInfo from all of the subtransactions, because it will contain a non-empty list of `subTxsInfo`, which in itself will contain all different TxInfos for all of the subtransactions.

I guess maybe we are lost in terminology here. Who is:
* regular observer?
* and a batch observer?

In my understanding both of those refer to the top level transaction. Because regular observer is the terminology that I only first heard in the CIP-112 and Batch observer is the term that I saw first in this CIP.

So, if both refer to scripts running for the top level transaction, then they should have the same TxInfo. But if either one of them refers to sub transactions then, naturally, they will have different TxInfos

After thinking more about it I am more and more convinced that batch observers from the top level transaction, as I was mentioning in the meeting, need to go. If we do not allow for subtransaction builders to specify the argument to their required observer, then those scripts do become useless:
* First of all there would be no way for them to know if that script satisfied their transaction or some other transaction.
* Outcome of execution of any sensible script does and should depend on the input. If the user has no ability so supply that input to the script it is requiring, and this ability instead is given to the top level transaction builder, then that script can be manipulated in any way by the top level transaction builder. Which is no good.

The only kind of script would work safely with "batch observers" is the one, which does not depend on any input, which is a totally useless script. So, after giving it some thought. Batch observers is not sound approach and it must go. If my argument still doesn't make sense, I'll be happy to meet and we can discuss it in more detail

With that out of the way let's look at a concrete example as a pseudo code.

```haskell
tx =
  Body
  { inputs = [txIn0:FOO 5]
  , outputs = [txOut:ADA 5]
  , fee = 5 ADA
  , subTx =
     [ Tx
       { body = Body
         { inputs = [txIn00:FOO 95]
         , outpus = [txOut00:ADA 200]
         , requiredObservers = [scriptHash00]
         }
       , wits = Wits
         { redeemers = [ (RedeemerPtr ReqObserver 0, (_exUnits000, data000))
                       , (RedeemerPtr Spending 0, (exUnits001, data001))
                       ]
         }
       }
     , Tx
       { body = Body
         { inputs = [txIn10:ADA 210]
         , outpus = [txOut10:FOO 100]
         , requiredObservers = [scriptHash10]
         }
       , wits = Wits
         { redeemers = [ (RedeemerPtr ReqObserver 0, (_exUnits100, data100))
                       ]
         }
       }
     ]
  , observers = [scriptHash0]
  }
  Wits {
   redeemers = [ (RedeemerPtr ReqObservers (0, 0)), (exUnits00, data00)
               , (RedeemerPtr ReqObservers (1, 0)), (exUnits10, data10)
               , (RedeemerPtr Spending 0, (exUnits0, data0))
               ]
  }
```
This is a nice example, since it captures couple of different uses cases:
* performing a swap 
* top level transaction builder is paying the fee with a multi asset (aka Babel fee)
* execution of scripts at all levels

So, if you look at the redeemers for the top level transaction there are three of them. Which means at the top level three independent script executions will occur and all of them will have the same `TxInfo`, which will contain all of the sub transactions.

Spending scripts, on the other hand, that specified in the subtransactions will have access only to their own TxInfo.
Note:
* that the underscore in `_exUnits000` and `_exUnits100`, it indicates those execution units have to be ignored, for obvious reasons.
* Both top level and sub transactions will have ability to supply data, eg, `scriptHash00` will receive both `data000` and `data00` as arguments. This will be a nice mechanism for top level transaction to communicate with the sub transaction  programmatically. Naturally `data00` could be ignored by the script, if it is not suppose to be used.

FYI I this is the kind of examples that CIP readers are looking for, rather than details on how we are going to enforce these with the ledger rules.

in order to create valid batches.

An additional high-level language may be beneficial to specify what the batch observer scripts 
actually require of the batch, as Plutus script constraints may be difficult to work with directly.

#### Light Client Service

The design outlined here can facilitate a trustless LC service.
Light clients may use the `corInputs` and `spendOuts` features to execute the following protocol.
LC is the light client, and SP is the service provider. We assume there exists a 
high-level language in which a light client can specify what they want a 
transaction constructed for them by the SP to do, e.g. pay `x` amount from my 
wallet to some other wallet, or mint some NFT. 

1. LC -> SP : a high-level specification `s` 
2. SP : construct a `tx` that 
- meets `s`, 
- includes one input `i` from SP's wallet
- includes a payment to SP for light-client services
3. SP : replace all inputs except `i` in `tx` with corresponding `spendOuts` from the UTxO set to make `tx'`
4. SP -> LC : `tx'`
5. LC : checks `tx'` meets `s`, signs it
6. LC -> SP : signed `tx'`
7. SP : constructs top-level `tx''` that provides the `corInputs` for `tx'`

> 8. SP : submits to Cardano the top-level `tx''` with sub-tx `tx'`

Author: > The part that LC needs is to know when someone payed that LC and when a transaction that was submitted on behalf of LC made it on-chain. I fail to see how this approach satisfies either one of these two requirements.

That would still have to be a separate mechanism, e.g. via a merkle proof in a Mithril snapshot once the Tx is included

> It might needs things like:
> current protocol parameters (eg. to calculate accurate fees or specify correct script integrtiy check)
> proposals, if submission of a vote is needed
> Current DReps if delegation is needed.

The goal of this approach is to co-construct a transaction between the LC and the SP in a way that provably satisfies the LC's constraints of what they want the transaction to do (while compensating SP for its services in that same transaction - either via payment, voting, or something else), but hides enough of the transaction data such that the LC cannot simply make a new transaction that doesnt compensate the SP for its services. 

The two extremes here are
- SP sends fully valid, unhashed and unencrypted transaction to LC (this may result in SP not getting paid)
- SP sends only the TxId of the transaction to LC (this may let SP trick LC into signing a tx that does something they dont want to do, like send all their $ to SP)

The hope is to find some middle ground, and maybe Nested Txs can help with this. The "overpaying the fee" problem can be addressed with nested Txs, because it can be specified as 0 for all LC transactions (and SP fee includes the transaction fee, as well as the script collateral)

We have done so by adding the following set of features :

1. transactions can be batched
- batch contains one top-level transaction, with multiple possible sub-transactions ;

2. the batch must be balanced, but individual transactions in it do not have to be, so 
- transactions in a single batch may pay each others' fees and minUTxOValue ;
- unbalanced transactions can be interpreted as swap offers, getting resolved within a complete batch ;

3. script data is shared across all transactions in a single batch 
- enables script deduplication ;

4. batch observer scripts enable sub-transactions to specify properties required of batches in which they may be included 
- design inspired by script observers ;
- represents all possible intents that do not require ledger rule relaxation ;

5. the top-level transaction must provide collateral for all scripts in its sub-transactions 
- sub-transactions are not required to cover their own collateral ;

6. a new intent we call "spend-by-output", wherein a sub-transaction may specify outputs it 
- intends to spend, and the top-level transaction specifies the inputs that point to those outputs in the UTxO set ;

> - example of possible additional supported intents (see CPS-15)

Reviewer 6: I am really sorry, but I have to push back on this feature, since it is fairly complex and is unclear to me. Which is a really good indicator that it needs its own CIP.
Nested transactions in itself is already a complex beast, we don't want to bloat it more than necessary. Since we'll have to implemented, I would for us to keep our sanity.

FYI, it would be nice to have a quote or title of the section from CPS-15 that talks about this use case.

In the case a set of sub-transactions is not fully balanced, and the aggregator is interested 
in participating in the exchange offers within the batch, it would have to include its own 
liquidity pool as one of the inputs 
in the top-level transaction, and output any extra tokens to its updated liquidity pool UTxO.  

### Babel fees

Babel fees is a specific subtype of the first usecase where the missing assets are necessarily
a quantity of Ada. Usually, it also implies that no Ada is contained in the inputs
of such a transaction, and the missing Ada goes towards paying transaction fees.

### DApp fee and min-UTxO sponsorship

DApp may run a Babel service which only cares only about transactions interacting with a 
certain DApp, discarding all other incoming transactions immediately. This service 
may include calculating ExUnits, paying fees/minUTxO for, and providing collateral for such 
underspecified transactions. 

### Comparison with Other Designs

### CIP-0131 "[Transaction Swaps](https://github.com/cardano-foundation/CIPs/pull/880)"

Transaction swaps achieve almost exactly the same goals as this CIP. The main differences 
are :

> 1. The top-level transaction contains subTxs`, which 

Reviewer 6: ```suggestion
1. The top-level transaction contains `subTxs`, which 
```

are lists of transactions. They are processed as individual transactions, alongside the top-level tx. As a result, 

- the `TxIn` of each output can be computed for each output underspecified transaction  
at the time of construction 

- the `TxId` of the transaction for each entry in the UTxO set is necessarily 

> signed bt all the keys required by that transaction

Reviewer 6: ```suggestion
signed by all the keys required by that transaction
```

- the input to (and therefore output of) each script is predictable at the time of (underspecified) tx construction, 
so that it is possible to compute required `ExUnits` for each of the scripts run by every sub-transaction 
at the time of construction

2. Contracts can only view data in the transaction running them, except in the case 
that a top-level tx is running a batch-observer script

3. In this design, script outputs (except of batch-observers) can be computed by any Babel service receiving an incoming 
underspecified transaction (as well as the author of the transaction), and they do not depend on the top-level transaction.
This makes constructing (and determining whether it is even possible to construct) 
a valid top-level transactions more straightforward. 

### CIP-0118 "Validation Zones-implicit"

The key difference between the Validation Zones-implicit design and this design is that 

> this design does not require as many (any?) major changes to be made to the operation 

Reviewer 6: ```suggestion
this design does not require any changes to be made to the operation 
```
